pydantic/_internal/_model_construction.py

```diff
@@ -388,19 +390,19 @@ def inspect_namespace(  # noqa C901
     return private_attributes
 
 
-def set_default_hash_func(namespace: dict[str, Any], bases: tuple[type[Any], ...]) -> None:
-    if '__hash__' in namespace:
-        return
-
+def set_default_hash_func(cls: type[BaseModel], bases: tuple[type[Any], ...]) -> None:
     base_hash_func = get_attribute_from_bases(bases, '__hash__')
     if base_hash_func in {None, object.__hash__}:
         # If `__hash__` is None _or_ `object.__hash__`, we generate a hash function.
         # It will be `None` if not overridden from BaseModel, but may be `object.__hash__` if there is another
         # parent class earlier in the bases which doesn't override `__hash__` (e.g. `typing.Generic`).
+
+        getter = operator.itemgetter(*cls.model_fields.keys()) if cls.model_fields else lambda _: None
+
         def hash_func(self: Any) -> int:
-            return hash(self.__class__) + hash(tuple(self.__dict__.values()))
+            return hash(self.__class__) + hash(getter(self.__dict__))
```

Comment:
Is this faster than just doing
```python
        def hash_func(self: Any) -> int:
            return hash(self.__class__) + hash(tuple(self.__dict__.get(k) for k in cls.model_fields))
```
I'd prefer to avoid adding the `operator` import if possible (as we have generally been trying to reduce the number of imports pydantic triggers, which increase startup time), but if it's noticeably faster then maybe it makes sense. I think if you can't see a significant and repeatable difference, better to not use it

Reply:
Doh I see there was a whole conversation about this and confirmation that it is significantly faster. 👍

Reply:
I still think we should confirm with @samuelcolvin though given the point about wanting to reduce import times.

Reply:
I see you propose using `.get()` instead of `[]`. Do you know whether it is safe to use `itemgetter()`, which doesn't have a default and thus will raise an error if the key is not found?

Said differently, are there cases where some keys from `cls.model_fields` are not present on the instance?

Reply:
> Said differently, are there cases where some keys from cls.model_fields are not present on the instance?

I tried to find ways that this might happen, which led me to https://github.com/pydantic/pydantic/issues/7784. I couldn't think of anything else but I don't actually know pydantic that deeply so I could easily be missing something. But my guess is that if it's possible then it's enough of a red flag that raising an error here would actually be a good thing.